Metadata Declaration in Core

DavidPoliakoff opened the pull request:
Currently looking for a review of this mechanism for declaring and printing metadata making sure that a code change like this wouldn't perturb anything, it needs more work. A sample output after the change, if you call print_configuration

```
Kokkos Version: 3.3.0
Compiler:
KOKKOS_COMPILER_GNU: 830
Architecture:
Default Device: N6Kokkos6SerialE
KOKKOS_ENABLE_ISA_KNC: no
KOKKOS_ENABLE_ISA_POWERPCLE: no
KOKKOS_ENABLE_ISA_X86_64: no
Atomics:
KOKKOS_ENABLE_GNU_ATOMICS: no
KOKKOS_ENABLE_INTEL_ATOMICS: no
KOKKOS_ENABLE_WINDOWS_ATOMICS: no
```

DavidPoliakoff commented:
Confirming that this can play nicely with the Tools infrastructure. So from my side, this is what I need

DavidPoliakoff commented:
Retest this please

dalg24 commented on cmake/kokkos_tribits.cmake:
```diff
@@ -277,6 +277,18 @@ MACRO(KOKKOS_CONFIGURE_CORE)
+       OUTPUT_VARIABLE KOKKOS_GIT_REVISION
+       WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+       RESULT_VARIABLE _RET)
+   if(NOT _RET EQUAL 0)
```
What can cause non-zero return value?

DavidPoliakoff replied:
If the copy of Kokkos doesn't come from a git repo

DavidPoliakoff replied:
Solved by deferring the feature

dalg24 commented on cmake/KokkosCore_config.h.in:
```diff
@@ -10,7 +10,7 @@
 // KOKKOS_VERSION / 10000 is the major version
 #cmakedefine KOKKOS_VERSION @KOKKOS_VERSION@
-
+#cmakedefine KOKKOS_GIT_REVISION @KOKKOS_GIT_REVISION@
```
```suggestion
#cmakedefine KOKKOS_GIT_REVISION @KOKKOS_GIT_REVISION@

```

DavidPoliakoff replied:
Solved by deferring the feature

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,206 @@ void initialize_profiling(const InitArguments& args) {
     std::exit(EXIT_SUCCESS);
   }
   Kokkos::Tools::parseArgs(args.tool_args);
+  for (const auto& category_values : Kokkos::Impl::metadata_map) {
```
```suggestion
  for (const auto& category_value : Kokkos::Impl::metadata_map) {
```

DavidPoliakoff replied:
0d662fbb

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,206 @@ void initialize_profiling(const InitArguments& args) {
   Kokkos::Tools::parseArgs(args.tool_args);
+  for (const auto& category_values : Kokkos::Impl::metadata_map) {
+    for (const auto& key_value : category_values.second) {
+      Kokkos::Tools::declareMetadata(key_value.first, key_value.second);
```
So you just throw away the "category" info?

DavidPoliakoff replied:
In declaring metadata to tools, yes. The category exists only to make metadata printing behave the same way

masterleinad commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,206 @@ void initialize_profiling(const InitArguments& args) {
+  }
 }
-
+#define KOKKOS_METADATA_QUOTE_HELPER(x) #x
+#define KOKKOS_METADATA_QUOTE(x) KOKKOS_METADATA_QUOTE_HELPER(x)
```
I would prefer defining
```
inline std::string kokkos_git_revision() { return "@KOKKOS_GIT_REVISION@"; }
```
in `cmake/KokkosCore_config.h.in` (or a separate file that we configure with `CMake`) over all these preprocessor defines.

DavidPoliakoff replied:
Me too. I was copying the other things, but if you're game for this, I'm more than happy to make the change

DavidPoliakoff replied:
We deferred this feature, based on Damien's comments. When we revisit, I'll likely use this suggestion, thanks

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -865,180 +1074,26 @@ void finalize_all() {
+void print_helper(std::ostringstream& out,
+                  const std::map<std::string, std::string>& print_me) {
+  for (const auto& kv : print_me) {
+    out << kv.first << ": " << kv.second << std::endl;
```
```suggestion
    out << kv.first << ": " << kv.second << '\n';
```

DavidPoliakoff replied:
577c2f0

masterleinad commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,206 @@ void initialize_profiling(const InitArguments& args) {
 void pre_initialize_internal(const InitArguments& args) {
   if (args.disable_warnings) g_show_warnings = false;
   if (args.tune_internals) g_tune_internals = true;
+  std::stringstream str_builder;
+  str_builder << KOKKOS_VERSION / 10000 << "." << (KOKKOS_VERSION % 10000) / 100
+              << "." << KOKKOS_VERSION % 100;
```
Maybe, refactor this into a separate function as well?

DavidPoliakoff replied:
I tried to make minimal changes, but sure

DavidPoliakoff replied:
577c2f0

masterleinad commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,206 @@ void initialize_profiling(const InitArguments& args) {
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_INTEL",
+                                 std::to_string(KOKKOS_COMPILER_INTEL));
+  declare_configuration_metadata("tools_only", "compiler_family", "intel");
+#endif
```
What about MSVC?

DavidPoliakoff replied:
I thought this was just a copy of the old version. If we want to add MSVC, I'd prefer it be in another PR

masterleinad replied:
Hmmm... I get
```bash
$ git grep "#define KOKKOS_COMPILER" | sort -u 
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_APPLECC __APPLE_CC__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_CLANG \
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_CRAYC _CRAYC
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_GNU \
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_IBM __IBMC__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_IBM __IBMCPP__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_IBM __ibmxl_vrm__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __ECC
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __ICC
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __INTEL_COMPILER
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __INTEL_LLVM_COMPILER
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_MSVC _MSC_VER
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_NVCC __NVCC__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_PGI \
```
so `MSVCC` really is the only one missing here and I would prefer to add it in this pull request.

DavidPoliakoff replied:
Do we want to add it to print_configuration, though? In the current implementation, print_configuration in Core doesn't say anything about MSVC. So we have three options

1) Change and add it to the printed metadata
2) Only add it to the declared metadata
3) Don't add it at all

I have no preferences among the three, it sounds like you prefer (1), but just to check, would you prefer (1), or (2)

DavidPoliakoff replied:
Talked in Slack, implemented (1) via 3e95a906

masterleinad commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,206 @@ void initialize_profiling(const InitArguments& args) {
+#else
+  declare_configuration_metadata("options", "KOKKOS_ENABLE_LIBRT", "no");
+#endif
+#ifdef KOKKOS_ENABLE_MPI
+  declare_configuration_metadata("options", "KOKKOS_ENABLE_MPI", "yes");
+#else
+  declare_configuration_metadata("options", "KOKKOS_ENABLE_MPI", "no");
+#endif
+  declare_configuration_metadata("architecture", "Default Device",
+                                 typeid(Kokkos::DefaultExecutionSpace).name());
```
This not a CMake option. We should likely remove or add it.

DavidPoliakoff replied:
This is being preserved from how the code used to work. It looks really dumb to me, but we used to print out this value in print_configuration. I'd prefer not to change this if we don't need to in this PR, we could do a follow-on

masterleinad replied:
Yes, absolutely. I didn't mean that you need to change anything here.

dalg24 commented on core/unit_test/CMakeLists.txt:
```diff
@@ -590,7 +590,7 @@ KOKKOS_ADD_ADVANCED_TEST( UnitTest_PushFinalizeHook_terminate
       TOOL kokkosprinter-tool
       ARGS --kokkos-tools-args="-c test delimit"
-      PASS_REGULAR_EXPRESSION "kokkosp_init_library::kokkosp_parse_args:4:KokkosCore_ProfilingAllCalls:-c:test:delimit::kokkosp_allocate_data:${MEMSPACE_REGEX}:source:${ADDRESS_REGEX}:40::kokkosp_begin_parallel_for:Kokkos::View::initialization [[]source]:0:0::kokkosp_end_parallel_for:0::kokkosp_allocate_data:${MEMSPACE_REGEX}:destination:${ADDRESS_REGEX}:40::kokkosp_begin_parallel_for:Kokkos::Vie ...
+      PASS_REGULAR_EXPRESSION "kokkosp_init_library::kokkosp_parse_args:4:KokkosCore_ProfilingAllCalls:-c:test:delimit::.*::kokkosp_allocate_data:${MEMSPACE_REGEX}:source:${ADDRESS_REGEX}:40::kokkosp_begin_parallel_for:Kokkos::View::initialization [[]source]:0:0::kokkosp_end_parallel_for:0::kokkosp_allocate_data:${MEMSPACE_REGEX}:destination:${ADDRESS_REGEX}:40::kokkosp_begin_parallel_for:Kokkos: ...
```
Unclear to me why this was needed update

DavidPoliakoff replied:
Because Kokkos::initialize now declares a boatload of metadata

dalg24 replied:
Oh I totally forgot that we had loaded a callback for the metadata declaration.

dalg24 reviewed (commented):
I would prefer if you tackled the recording the git commit hash in a separate PR

DavidPoliakoff commented:
@masterleinad @dalg24 I think I've addressed the comments. I deferred the Git SHA stuff for a future PR, and _almost_ everything else I just implemented as suggested. The only exception is that I'm not adding the machinery to print MSVC, as we didn't used to print that in our print_configuration machinery

masterleinad commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -355,7 +355,11 @@ void pre_initialize_internal(const InitArguments& args) {
 #endif
-
+#ifdef KOKKOS_COMPILER_MSVC
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_PGI",
```
```suggestion
  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_MSVC",
```

DavidPoliakoff replied:
OOOOOOOF

DavidPoliakoff replied:
Good catch. Yikes

dalg24 commented on core/src/Kokkos_Core.hpp:
```diff
@@ -144,6 +144,10 @@ void pre_initialize(const InitArguments& args);
 
 void post_initialize(const InitArguments& args);
 
+void declare_configuration_metadata(const std::string& category,
+                                    const std::string& key,
+                                    const std::string& value);
```
Why do you need the declaration in `Kokkos_Core.hpp`?  Where else than `Kokkos_Core.cpp` would we use?

DavidPoliakoff replied:
This could be used in backends, the CUDA backend will likely declare configuration metadata as well

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -865,180 +1080,26 @@ void finalize_all() {
-#else
-  msg << "no" << std::endl;
-#endif
+  print_helper(msg, Kokkos::Impl::metadata_map["atomics"]);
```
Not printing `"options"`?

DavidPoliakoff replied:
Or memory, weird, I'll change that

DavidPoliakoff replied:
In 8721e87

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,212 @@ void initialize_profiling(const InitArguments& args) {
+#else
+  declare_configuration_metadata("options", "KOKKOS_ENABLE_MPI", "no");
+#endif
+  declare_configuration_metadata("architecture", "Default Device",
```
Why is this part of `"architecture"`?

DavidPoliakoff replied:
This is part of architecture because that's where it printed in print_configuration before. We can change that, but the underlying goal in this PR was to not shake up the printing as it happened before

DavidPoliakoff commented on cmake/KokkosCore_config.h.in:
```diff
@@ -10,7 +10,6 @@
 // KOKKOS_VERSION / 100 % 100 is the minor version
 // KOKKOS_VERSION / 10000 is the major version
 #cmakedefine KOKKOS_VERSION @KOKKOS_VERSION@
-
```
In 8721e87

DavidPoliakoff commented:
@dalg24, remaining are questions about making the default device part of the architecture (which it was before, but which I'm not opposed to changing), and whether we want to expose the `declare_configuration_metadata` symbol for use in other places

DavidPoliakoff commented:
Retest this please

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -282,11 +295,212 @@ void initialize_profiling(const InitArguments& args) {
+              << "." << version_number % 100;
+  return str_builder.str();
+}
+#define KOKKOS_METADATA_QUOTE_HELPER(x) #x
+#define KOKKOS_METADATA_QUOTE(x) KOKKOS_METADATA_QUOTE_HELPER(x)
```
Unused now that we pulled the git commit hash out

DavidPoliakoff replied:
7230f06

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -81,6 +81,19 @@ std::stack<hook_function_type, std::list<hook_function_type>> finalize_hooks;
+using metadata_category_type = std::string;
+using metadata_key_type      = std::string;
+using metadata_value_type    = std::string;
+
```
Comment that category is only used to group related key-value pairs when printing the configuration and that the information is not passed to the tools.

DavidPoliakoff replied:
7230f06

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -81,6 +81,19 @@ std::stack<hook_function_type, std::list<hook_function_type>> finalize_hooks;
+using metadata_category_type = std::string;
+using metadata_key_type      = std::string;
+using metadata_value_type    = std::string;
+
```
Comment that category is only used to group related key-value pairs when printing the configuration and that the information is not passed to the tools.

DavidPoliakoff replied:
7230f06

dalg24 reviewed (changes requested):
Add the comment and remove the unused code.

Also I am curious whether you considered generating the tedious bit
```C++
#ifdef KOKKOS_ENABLE_FOO
declare_configuration_metadata("options", "KOKKOS_ENABLE_FOO", "yes");
#else
declare_configuration_metadata("options", "KOKKOS_ENABLE_FOO", "no");
#endif
```
with CMake.  (NOT asking you to change)

DavidPoliakoff commented:
@dalg24, I copied that tedious bit from our previous implementation. That implementation doesn't look good to me, but every time we change something to "something that's exactly the same" I find out that we broke something somehow

DavidPoliakoff commented:
@dalg24, believe I have addressed the comments

dalg24 reviewed (approved):
(Need to clean up history)

DavidPoliakoff commented:
@dalg24: done
